### PR TITLE
[4.1.x][6954] XB does not show up added repeating group item and need a browser refresh

### PR DESCRIPTION
**Pull request by jvega190**
https://github.com/craftercms/craftercms/issues/6954

### CODE REVIEW

File: ui/guest/src/contentController.ts [hunk 1]
```diff
@@ -385,6 +385,16 @@ export function insertItem(modelId: string, fieldId: string, index: number | str
     }
   });
 
+  const currentItems = models[modelId][fieldId];
+  const newItems = index ? currentItems.splice(index, 0, instance) : [...(currentItems ?? []), instance];
```
Comment (rart) [inline ui/guest/src/contentController.ts]: Don't modify/mutate the existing array instance.

**Review by rart — CHANGES_REQUESTED** [review]

File: ui/guest/src/contentController.ts [hunk 2]
```diff
@@ -385,6 +385,16 @@ export function insertItem(modelId: string, fieldId: string, index: number | str
     }
   });
 
+  const currentItems = (models[modelId][fieldId] ?? []).slice();
```
Comment (rart) [inline ui/guest/src/contentController.ts]:
```suggestion
  const currentItems = [...(models[modelId][fieldId] ?? [])];
```

**Review by rart — CHANGES_REQUESTED** [review]

File: ui/guest/src/contentController.ts [hunk 3]
```diff
@@ -385,6 +385,16 @@ export function insertItem(modelId: string, fieldId: string, index: number | str
...
 
+  const currentItems = [...(models[modelId][fieldId] ?? [])];
+  const newItems = index ? currentItems.splice(index, 0, instance) : [...currentItems, instance];
+  models$.next({
+    ...models,
+    [modelId]: {
+      ...models[modelId],
+      [fieldId]: newItems
+    }
+  });
```
Comment (rart) [inline ui/guest/src/contentController.ts]: Create a nested fields/indexes example (e.g. repeat_o.itemSelector_o, 0.1) and see if this works. My guess if we need to accommodate the dotted notation by using the utility functions we have.

**Review by rart — CHANGES_REQUESTED** [review]

File: ui/guest/src/utils/ice.ts [hunk 4]
```diff
@@ -79,7 +79,7 @@ export function getCollectionWithoutItemAtIndex(collection: string[], index: str
   return collection.slice(0, parsedIndex).concat(collection.slice(parsedIndex + 1));
 }
 
-export function getCollection(model: ContentInstance, fieldId: string, index: string | number): string[] {
+export function getCollection(model: ContentInstance, fieldId: string, index: string | number) {
```
Comment (jvega190) [inline ui/guest/src/utils/ice.ts]: @rart will ping you to ask a couple of things about this

File: ui/guest/src/contentController.ts [hunk 5]
```diff
@@ -379,12 +380,37 @@ export function duplicateItem(modelId: string, fieldId: string, index: number |
 export function insertItem(modelId: string, fieldId: string, index: number | string, contentType: ContentType): void {
   const instance: InstanceRecord = {};
   const models = getCachedModels();
-  Object.entries(contentType.fields[fieldId].fields).forEach(([id, field]) => {
+  const fields = fieldId.split('.');
+  const field = fields.reduce((acc, field) => acc[field].fields, contentType.fields);
```
Comment (rart) [inline ui/guest/src/contentController.ts]:
```suggestion
  const fieldPathArray = fieldId.split('.');
  const repeatFields = fields.reduce((acc, field) => acc[field].fields, contentType.fields);
```

File: ui/guest/src/contentController.ts [hunk 6]
```diff
@@ -379,12 +380,37 @@ export function duplicateItem(modelId: string, fieldId: string, index: number |
...
+  Object.entries(field).forEach(([id, field]) => {
     if (!systemProps.includes(field.id)) {
       instance[id] = getDefaultValue(field);
     }
   });
 
+  const indexPieces = index?.toString().split('.');
+  const concatFieldId = !isSimple(fieldId)
+    ? mergeArraysAlternatively(fieldId.split('.'), indexPieces).join('.')
+    : fieldId;
```
Comment (rart) [inline ui/guest/src/contentController.ts]:
```suggestion
  const fieldPathWithIndexes = isSimple(fieldId)
    ? fieldId
    : mergeArraysAlternatively(fieldId.split('.'), indexPieces).join('.');
```

File: ui/guest/src/contentController.ts [hunk 7]
```diff
@@ -379,12 +380,37 @@ export function duplicateItem(modelId: string, fieldId: string, index: number |
...
+  const currentItems = getCollection(models[modelId], concatFieldId, index) ?? [];
```
Comment (rart) [inline ui/guest/src/contentController.ts]: Strange to send the field with indexes and an index also. Pls check

File: ui/guest/src/contentController.ts [hunk 8]
```diff
@@ -379,12 +380,37 @@ export function duplicateItem(modelId: string, fieldId: string, index: number |
...
+  if (nou(index) || indexPieces.length < fields.length) {
+    // Index doesn't exist for current field subsection (when not simple field), or index doesn't exist at all, insert at the end
+    newItems = [...currentItems, instance];
+  } else {
+    // If index is complex, we already have the set of items from the model, now we need the latter part of the index
+    // to insert the new instance at the desired position.
+    let targetIndex = isSimple(index) ? index : popPiece(index as string);
+    newItems = currentItems.splice(targetIndex as number, 0, instance);
+  }
+  const model = setCollection(models[modelId], fieldId, index, newItems);
```
Comment (rart) [inline ui/guest/src/contentController.ts]: Do not mutate the existing instance inside of models.

**Review by rart — CHANGES_REQUESTED** [review]

File: ui/guest/src/contentController.ts (hunk 8 above)
Comment (rart) [reply ui/guest/src/contentController.ts]: The additional complexity here is that shouldn't mutate the inner object target of the collection being set

**Review by rart — APPROVED** [review]

**Pull request merged**